File: leetcode15/leetcode15.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
    public:
        vector<vector<int>> threeSum(vector<int>& nums) {
            vector<vector<int>> res;
            sort(nums.begin(), nums.end());
            for (int i = 0; i < nums.size(); i++){
                if (nums[i] > 0) return res;
                if (i > 0 && nums[i] == nums[i - 1]) continue;
                int left = i + 1;
                int right = nums.size() - 1;
                while(right > left){
                    int sum = nums[i] + nums[left] + nums[right];
                    if(sum > 0) right --;
                    if(sum < 0) left ++;
                    if(sum == 0){
                        res.push_back(vector<int>{nums[i], nums[left], nums[right]});
                        while(right > left && nums[left] == nums[left + 1]) left ++;
                        while(right > left && nums[right] == nums[right - 1]) right --;
                        left ++;
                        right --;
                    }
                }
            }
            return res;
        }
    };

File: leetcode15/leetcode15.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for i in range(len(nums)):
            if nums[i] > 0:
                return res
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            left = i + 1
            right = len(nums) - 1
            while right > left:
                sum_ = nums[i] + nums[left] + nums[right]
                if sum_ < 0:
                    left += 1
                if sum_ > 0:
                    right -= 1
                if sum_ == 0:
                    res.append([nums[i], nums[left], nums[right]])

                    while right > left and nums[left] == nums[left + 1]:
                        left += 1
                    while right > left and nums[right] == nums[right - 1]:
                        right -= 1
                    left += 1
                    right -= 1
        return res

